File: expressiveWords/expressiveWords.java
class Solution {
    public int expressiveWords(String S, String[] words) {
        if (S == null || words == null) { 
            return 0;
        }
        int res = 0;
        for (String word : words) {
            if (stretchy(S, word)) {
                res++;
            }
        }
        return res;
    }
    
    private boolean stretchy(String S, String word) {
        if (word == null) {
            return false;
        }
        
        int i = 0;
        int j = 0;
        while (i < S.length() && j < word.length()) {
            if (S.charAt(i) == word.charAt(j)) {
                int repS = getRepeated(S, i);
                int repW = getRepeated(word, j);
                //System.out.println(repS + " " + repW);
                if (repS < repW || (repS != repW && repS < 3)) {
                    return false;
                }
                i += repS;
                j += repW;
            } else {
                return false;
            }
        }
        return i == S.length() && j == word.length();
    }
    
    private int getRepeated(String S, int index) {
        if (S == null) {
            return 0;
        }
        int i = index + 1;
        int cnt = 1;
        while (i < S.length()) {
            if (S.charAt(i) == S.charAt(i - 1)) {
                cnt++;
            } else {
                break;
            }
            i++;
        }
        return cnt;
    }
}

File: 809.expressiveWords/expressiveWords.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int expressiveWords(string S, vector<string>& words) {
        int res = 0;
        if (S.length() == 0 || words.size() == 0)
            return res;
        string compS = compress(S);
        //cout << compressS << endl;
        for (string& word : words) {
            string compW = compress(word);
            if (compS.length() != compW.length())
                continue;
            bool couldExpend = true;
            for (int i = 0; i < compS.length(); i += 2) {
                if (compS[i] != compW[i] || compS[i + 1] < compW[i + 1]) {
                    couldExpend = false;
                    break;
                }
                if (compS[i + 1] - '0' < 3 && compS[i + 1] != compW[i + 1]) {
                    couldExpend = false;
                    break;
                }
            }
            if (couldExpend)
                res++;
        }
        return res;
    }
    string compress(string s) {
        string res;
        if (s.length() == 0)
            return res;
        int count = 1;
        res = s[0];
        for (int i = 1; i < s.length(); i++) {
            if (s[i] == s[i - 1]) {
                count++;
            } else {
                res += to_string(count) + s[i];
                count = 1;
            }
        }
        res += to_string(count);
        return res;
    }
};

int main() {
  string S = "heeellooo";
  vector<string> words = {"hello", "hi", "helo"};
  Solution s;
  cout << s.expressiveWords(S, words) << endl;
  return 0;
}
